builds.js

```diff
@@ -9,112 +12,327 @@ const environments = new Map([
+   * @param {Object} [options.buildCache] An options object for caching build information
+   * @param {boolean} [options.buildCache.enabled] True if you want to use a cache of build information,
+   * False if you want to hit the service every time
+   * @param {number} [options.buildCache.refreshLimit] The number of cache entries that can be refreshed simultaneously.
+   * @param {number} [options.buildCache.refreshInterval] An interval in ms on which the cache should be refreshed.
+   * If the refresh takes longer than the interval, intervals will be skipped. A non-positive number results in the
+   * cache _never_ being refreshed
+   * @public
+   */
+  constructor(warehouse, options) {
+    this.warehouse = warehouse;
+    this._cacheRefreshIntervalId = null;
 
-/**
- * Get build information.
- *
- * @param {Object} params Parameters that specify pkg, env, version and/or locale.
- * @param {Function} fn Completion callback.
- * @returns {Warehouse} fluent interface.
- * @api public
- */
-Builds.prototype.get = function get(params, fn) {
-  const env = environments.get(params.environment || params.env) || 'dev';
-  const version = params.v || params.version;
-  const meta = params.meta ? 'meta' : null;
-  const locale = params.locale || 'en-US';
-  let pkg = params.package || params.pkg;
-
-  if (!pkg || !env) {
-    return fn(
-      new Error('invalid parameters supplied, missing `pkg` or `env`')
-    );
+    options = options || {};
+    if (options.buildCache && options.buildCache.enabled) {
+      this._refreshCache = this._refreshCache.bind(this);
+
+      /**
+       * @member {Map} cache The build cache
+       * @private
+       */
+      this._cache = new Map();
+
+      /**
+       * @member {number} _cacheRefreshLimit The number of cache entries that can be simultaneously refreshed
+       * @private
+       */
+      this._cacheRefreshLimit = options.buildCache.refreshLimit || defaultCacheRefreshLimit;
+
+      /**
+       * @member {number} _cacheRefreshInterval The interval in ms on which the cache is refreshed refreshed
+       * @private
+       */
+      this._cacheRefreshInterval = (options.buildCache.refreshInterval || options.buildCache.refreshInterval === 0) ?
+        options.buildCache.refreshInterval :
+        defaultCacheRefreshInterval;
+      this.resumeCacheRefresh();
+    }
   }
 
-  //
-  // Handle scoped packages
-  //
-  pkg = encodeURIComponent(pkg);
+  /**
+   * Clears the build cache if one exists
+   *
+   * @public
+   */
+  clearCache() {
+    if (this._cache) {
+      this._cache.clear();
+    }
+  }
 
-  debug('Build metadata: pkg = %s, env = %s, version = %s', pkg, env, version);
-  return this.warehouse.send(
-    ['builds'].concat(pkg, env, version, meta).filter(Boolean).join('/'),
-    { query: { locale }},
-    fn
-  );
-};
+  /**
+   * Allows you to resume the process of refreshing the build cache
+   *
+   * @public
+   */
+  resumeCacheRefresh() {
```

Comment:
Part of me really wants to have this decoupled into a separate `cache` class that has this refresh ability rather than it being contained within the `Builds` class but I'm ok leaving that as a future refactor if its too much work at the moment.

Reply:
I was actually thinking about this last night after I sent this out.  I think I'm going to do this anyway.  It'd also provide a way to bring your own cache if necessary and gets a lot of the nastiness out of the Builds class itself.